pkg/graveler/settings/manager.go

```diff
@@ -70,39 +68,64 @@ func (m *Manager) Save(ctx context.Context, repository *graveler.RepositoryRecor
 // The setting is persisted only if the current version of the setting matches the given checksum.
-func (m *Manager) SaveIf(ctx context.Context, repository *graveler.RepositoryRecord, key string, setting proto.Message, lastKnownChecksum string) error {
+// If lastKnownChecksum is nil, the setting is persisted only if it does not exist.
+func (m *Manager) SaveIf(ctx context.Context, repository *graveler.RepositoryRecord, key string, setting proto.Message, lastKnownChecksum *string) error {
 	logSetting(logging.FromContext(ctx), repository.RepositoryID, key, setting, "saving repository-level setting")
-	decodedChecksum, err := base64.StdEncoding.DecodeString(lastKnownChecksum)
-	if err != nil {
-		return fmt.Errorf("decode checksum: %w", err)
+	valueWithPredicate, err := m.store.Get(ctx, []byte(graveler.RepoPartition(repository)), []byte(graveler.SettingsPath(key)))
+	if err != nil && !errors.Is(err, kv.ErrNotFound) {
+		return err
+	}
+	var currentChecksum *string
+	var currentPredicate kv.Predicate
+	if valueWithPredicate != nil {
+		if valueWithPredicate.Value != nil {
+			currentChecksum, err = computeChecksum(valueWithPredicate.Value)
+		}
+		if err != nil {
+			return err
+		}
+		currentPredicate = valueWithPredicate.Predicate
+	}
+	if swag.StringValue(currentChecksum) != swag.StringValue(lastKnownChecksum) {
+		return graveler.ErrPreconditionFailed
+	}
+	err = kv.SetMsgIf(ctx, m.store, graveler.RepoPartition(repository), []byte(graveler.SettingsPath(key)), setting, currentPredicate)
+	if err != nil && errors.Is(err, kv.ErrPredicateFailed) {
+		return graveler.ErrPreconditionFailed
 	}
-	return kv.SetMsgIf(ctx, m.store, graveler.RepoPartition(repository), []byte(graveler.SettingsPath(key)), setting, decodedChecksum)
+	return err
 }
 
-func (m *Manager) getWithPredicate(ctx context.Context, repo *graveler.RepositoryRecord, key string, data proto.Message) (kv.Predicate, error) {
-	pred, err := kv.GetMsg(ctx, m.store, graveler.RepoPartition(repo), []byte(graveler.SettingsPath(key)), data)
+func computeChecksum(value []byte) (*string, error) {
+	h := sha256.New()
+	_, err := h.Write(value)
 	if err != nil {
-		if errors.Is(err, kv.ErrNotFound) {
-			err = graveler.ErrNotFound
-		}
 		return nil, err
 	}
-	return pred, nil
+	return swag.String(hex.EncodeToString(h.Sum(nil))), nil
 }
 
 // GetLatest returns the latest setting under the given repository and key, without using the cache.
 // The returned checksum represents the version of the setting, and can be passed to SaveIf for conditional updates.
-func (m *Manager) GetLatest(ctx context.Context, repository *graveler.RepositoryRecord, key string, settingTemplate proto.Message) (proto.Message, string, error) {
-	data := settingTemplate.ProtoReflect().Interface()
-	pred, err := m.getWithPredicate(ctx, repository, key, data)
+func (m *Manager) GetLatest(ctx context.Context, repository *graveler.RepositoryRecord, key string, settingTemplate proto.Message) (proto.Message, *string, error) {
```

Comment:
Accept `protoreflect.ProtoMessage` as settingMsg and unmarshal the data into it instead of template+reflect+create and return.